Neos.ContentRepository.Core/Classes/ContentGraphFinder.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+/*
+ * This file is part of the Neos.ContentRepository package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+declare(strict_types=1);
+
+namespace Neos\ContentRepository\Core;
+
+use Neos\ContentRepository\Core\Projection\ContentGraph\ContentGraphInterface;
+use Neos\ContentRepository\Core\SharedModel\Exception\ContentStreamDoesNotExistYet;
+use Neos\ContentRepository\Core\SharedModel\Exception\WorkspaceDoesNotExist;
+use Neos\ContentRepository\Core\SharedModel\Workspace\ContentStreamId;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+
+/**
+ * A finder for ContentGraphInterface bound to contentStream/Workspace
+ *
+ * @see ContentGraphInterface
+ * @api only for read access during write operations and in services
+ */
+final class ContentGraphFinder implements ContentGraphFinderInterface
+{
+    /**
+     * @var array<string, ContentGraphInterface>
+     */
+    private array $contenGraphInstances = [];
+
+    public function __construct(
+        private readonly ContentGraphFactoryInterface $contentGraphFactory
+    ) {
+    }
+
+    /**
+     * @throws WorkspaceDoesNotExist if there is no workspace with the provided name
+     * @throws ContentStreamDoesNotExistYet if the provided workspace does not resolve to an existing content stream
+     */
+    public function fromWorkspaceName(WorkspaceName $workspaceName): ContentGraphInterface
+    {
+        if (isset($this->contenGraphInstances[$workspaceName->value])) {
+            return $this->contenGraphInstances[$workspaceName->value];
+        }
+
+        $this->contenGraphInstances[$workspaceName->value] = $this->contentGraphFactory->buildForWorkspace($workspaceName);
+        return $this->contenGraphInstances[$workspaceName->value];
+    }
+
+    /**
+     * @return ContentGraphInterface[]
+     */
+    public function getInstances(): array
+    {
+        return $this->contenGraphInstances;
+    }
+
+    public function reset(): void
+    {
+        $this->contenGraphInstances = [];
+    }
+
+    /**
+     * @param WorkspaceName $workspaceName
+     * @param ContentStreamId $contentStreamId
+     * @return ContentGraphInterface
+     * @internal
+     */
+    public function fromWorkspaceNameAndContentStreamId(WorkspaceName $workspaceName, ContentStreamId $contentStreamId): ContentGraphInterface
```

Comment:
So this is ONLY used in tests, and TBH again last minute addition because I didn't see an easy way out. We need to talk about this. The behat tests do some ugly stuff internally...

Reply:
Yes i know the test use nice reflection and stuff ^^ i started a conversation here about this https://github.com/neos/neos-development-collection/issues/4634